In [1]:
import pandas as pd

Import a csv file from the local drive into a DataFrame named df  

In [37]:
df = pd.read_csv('pokemon_data.csv')
df

#                   Name   Type 1  Type 2  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison  45      49       49   
1      2                Ivysaur    Grass  Poison  60      62       63   
2      3               Venusaur    Grass  Poison  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison  80     100      123   
4      4             Charmander     Fire     NaN  39      52       43   
..   ...                    ...      ...     ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark  80     160       60   
799  721              Volcanion     Fire   Water  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 12 columns]

Display the first/last five rows in df using head/tail respectively (NOTE: To print all the rows, use 'pd.options.display.max_rows = None' and then 'df')

In [38]:
df.head(5)

#                   Name Type 1  Type 2  HP  Attack  Defense  Sp. Atk  \
0  1              Bulbasaur  Grass  Poison  45      49       49       65   
1  2                Ivysaur  Grass  Poison  60      62       63       80   
2  3               Venusaur  Grass  Poison  80      82       83      100   
3  3  VenusaurMega Venusaur  Grass  Poison  80     100      123      122   
4  4             Charmander   Fire     NaN  39      52       43       60   

   Sp. Def  Speed  Generation  Legendary  
0       65     45           1      False  
1       80     60           1      False  
2      100     80           1      False  
3      120     80           1      False  
4       50     65           1      False

Find the total number of rows and columns in df

In [9]:
df.shape

(800, 12)

Display the names of all columns in df

In [19]:
df.columns

Index(['#', 'Name', 'Type 1', 'Type 2', 'HP', 'Attack', 'Defense', 'Sp. Atk',
       'Sp. Def', 'Speed', 'Generation', 'Legendary'],
      dtype='object')

Display the names of all rows in df

In [20]:
df.index

RangeIndex(start=0, stop=800, step=1)

Display the name of the fifth column

In [27]:
df.columns[5]

'Attack'

Select the row 2 from df using iloc (index begin from 0 by default)

In [5]:
df.iloc[2]

#                    3
Name          Venusaur
Type 1           Grass
Type 2          Poison
HP                  80
Attack              82
Defense             83
Sp. Atk            100
Sp. Def            100
Speed               80
Generation           1
Legendary        False
Name: 2, dtype: object

Select rows till 5 using iloc

In [6]:
df.iloc[:5]

#                   Name Type 1  Type 2  HP  Attack  Defense  Sp. Atk  \
0  1              Bulbasaur  Grass  Poison  45      49       49       65   
1  2                Ivysaur  Grass  Poison  60      62       63       80   
2  3               Venusaur  Grass  Poison  80      82       83      100   
3  3  VenusaurMega Venusaur  Grass  Poison  80     100      123      122   
4  4             Charmander   Fire     NaN  39      52       43       60   

   Sp. Def  Speed  Generation  Legendary  
0       65     45           1      False  
1       80     60           1      False  
2      100     80           1      False  
3      120     80           1      False  
4       50     65           1      False

Select rows from 100 till the end

In [7]:
df.iloc[100:]

#                 Name   Type 1  Type 2  HP  Attack  Defense  Sp. Atk  \
100   93              Haunter    Ghost  Poison  45      50       45      115   
101   94               Gengar    Ghost  Poison  60      65       60      130   
102   94    GengarMega Gengar    Ghost  Poison  60      65       80      170   
103   95                 Onix     Rock  Ground  35      45      160       30   
104   96              Drowzee  Psychic     NaN  60      48       45       43   
..   ...                  ...      ...     ...  ..     ...      ...      ...   
795  719              Diancie     Rock   Fairy  50     100      150      100   
796  719  DiancieMega Diancie     Rock   Fairy  50     160      110      160   
797  720  HoopaHoopa Confined  Psychic   Ghost  80     110       60      150   
798  720   HoopaHoopa Unbound  Psychic    Dark  80     160       60      170   
799  721            Volcanion     Fire   Water  80     110      120      130   

     Sp. Def  Speed  Generation  Legendary  
100       55     95           1      False  
101       75    110           1      False  
102       95    130           1      False  
103       45     70           1      False  
104       90     42           1      False  
..       ...    ...         ...        ...  
795      150     50           6       True  
796      110    110           6       True  
797      130     70           6       True  
798      130     80           6       True  
799       90     70           6       True  

[700 rows x 12 columns]

Select 5 rows from the end 

In [8]:
df.iloc[-5:]

#                 Name   Type 1 Type 2  HP  Attack  Defense  Sp. Atk  \
795  719              Diancie     Rock  Fairy  50     100      150      100   
796  719  DiancieMega Diancie     Rock  Fairy  50     160      110      160   
797  720  HoopaHoopa Confined  Psychic  Ghost  80     110       60      150   
798  720   HoopaHoopa Unbound  Psychic   Dark  80     160       60      170   
799  721            Volcanion     Fire  Water  80     110      120      130   

     Sp. Def  Speed  Generation  Legendary  
795      150     50           6       True  
796      110    110           6       True  
797      130     70           6       True  
798      130     80           6       True  
799       90     70           6       True

Select the value in row 2 and column 3 using iloc

In [10]:
df.iloc[2,3]

'Poison'

Select rows till 2 and columns till 4 using iloc

In [41]:
df.iloc[:2,:4]

#       Name Type 1  Type 2
0  1  Bulbasaur  Grass  Poison
1  2    Ivysaur  Grass  Poison

Select all rows and columns till 3

In [12]:
df.iloc[:,:3]

#                   Name   Type 1
0      1              Bulbasaur    Grass
1      2                Ivysaur    Grass
2      3               Venusaur    Grass
3      3  VenusaurMega Venusaur    Grass
4      4             Charmander     Fire
..   ...                    ...      ...
795  719                Diancie     Rock
796  719    DiancieMega Diancie     Rock
797  720    HoopaHoopa Confined  Psychic
798  720     HoopaHoopa Unbound  Psychic
799  721              Volcanion     Fire

[800 rows x 3 columns]

Selecting rows till 3 using loc (NOTE: Using df.iloc[:3,:] would not include row 3, whereas df.loc[:3,:] includes the third row)

In [87]:
df.loc[:3,:]

#                   Name Type 1  Type 2  HP  Attack  Defense  Sp. Atk  \
0  1              Bulbasaur  Grass  Poison  45      49       49       65   
1  2                Ivysaur  Grass  Poison  60      62       63       80   
2  3               Venusaur  Grass  Poison  80      82       83      100   
3  3  VenusaurMega Venusaur  Grass  Poison  80     100      123      122   

   Sp. Def  Speed  Generation  Legendary  New HP  New HP1  New HP2  New HP3  
0       65     45           1      False       0       50       45      NaN  
1       80     60           1      False       1       50       61      NaN  
2      100     80           1      False       2       50       82      NaN  
3      120     80           1      False       3       50       83      NaN

Select all values from a column named 'Name'

In [31]:
df['Name']

0                  Bulbasaur
1                    Ivysaur
2                   Venusaur
3      VenusaurMega Venusaur
4                 Charmander
               ...          
795                  Diancie
796      DiancieMega Diancie
797      HoopaHoopa Confined
798       HoopaHoopa Unbound
799                Volcanion
Name: Name, Length: 800, dtype: object

Select the value in row 30 in column 'Name'

In [33]:
df['Name'][30]

'Pikachu'

Select values from row 790 till end in 'Name' 

In [37]:
df['Name'][790:]

790                 Noibat
791                Noivern
792                Xerneas
793                Yveltal
794       Zygarde50% Forme
795                Diancie
796    DiancieMega Diancie
797    HoopaHoopa Confined
798     HoopaHoopa Unbound
799              Volcanion
Name: Name, dtype: object

Select three columns 'Name','Type 1' and 'HP'

In [42]:
df[['Name','Type 1','HP']]

Name   Type 1  HP
0                Bulbasaur    Grass  45
1                  Ivysaur    Grass  60
2                 Venusaur    Grass  80
3    VenusaurMega Venusaur    Grass  80
4               Charmander     Fire  39
..                     ...      ...  ..
795                Diancie     Rock  50
796    DiancieMega Diancie     Rock  50
797    HoopaHoopa Confined  Psychic  80
798     HoopaHoopa Unbound  Psychic  80
799              Volcanion     Fire  80

[800 rows x 3 columns]

Rename the column names by passing a dictionary. Rename is used to rename the row/column names whereas replace is used to rename the values inside the row/column. 

In [115]:
df.rename(columns = {'Name':'P Name','#':'Serial no'})

Serial no                 P Name   Type 1  Type 2  HP  Attack  Defense  \
0            1              Bulbasaur    Grass  Poison  45      49       49   
1            2                Ivysaur    Grass  Poison  60      62       63   
2            3               Venusaur    Grass  Poison  80      82       83   
3            3  VenusaurMega Venusaur    Grass  Poison  80     100      123   
4            4             Charmander     Fire     NaN  39      52       43   
..         ...                    ...      ...     ...  ..     ...      ...   
795        719                Diancie     Rock   Fairy  50     100      150   
796        719    DiancieMega Diancie     Rock   Fairy  50     160      110   
797        720    HoopaHoopa Confined  Psychic   Ghost  80     110       60   
798        720     HoopaHoopa Unbound  Psychic    Dark  80     160       60   
799        721              Volcanion     Fire   Water  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 12 columns]

Naming the x-axis as 'Serial no' and y-axis as 'Details' 

In [99]:
df.rename_axis('Serial no')
df.rename_axis('Details', axis = 1)

Details    #                   Name   Type 1  Type 2  HP  Attack  Defense  \
0          1              Bulbasaur    Grass  Poison  45      49       49   
1          2                Ivysaur    Grass  Poison  60      62       63   
2          3               Venusaur    Grass  Poison  80      82       83   
3          3  VenusaurMega Venusaur    Grass  Poison  80     100      123   
4          4             Charmander     Fire     NaN  39      52       43   
..       ...                    ...      ...     ...  ..     ...      ...   
795      719                Diancie     Rock   Fairy  50     100      150   
796      719    DiancieMega Diancie     Rock   Fairy  50     160      110   
797      720    HoopaHoopa Confined  Psychic   Ghost  80     110       60   
798      720     HoopaHoopa Unbound  Psychic    Dark  80     160       60   
799      721              Volcanion     Fire   Water  80     110      120   

Details  Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0             65       65     45           1      False  
1             80       80     60           1      False  
2            100      100     80           1      False  
3            122      120     80           1      False  
4             60       50     65           1      False  
..           ...      ...    ...         ...        ...  
795          100      150     50           6       True  
796          160      110    110           6       True  
797          150      130     70           6       True  
798          170      130     80           6       True  
799          130       90     70           6       True  

[800 rows x 12 columns]

Filter rows with the value 'Fire' in column 'Type 1' using loc and limit it to 5 using tail

In [51]:
df.loc[df['Type 1'] == 'Fire'].tail(5)

#         Name Type 1  Type 2  HP  Attack  Defense  Sp. Atk  Sp. Def  \
730  662  Fletchinder   Fire  Flying  62      73       55       56       52   
731  663   Talonflame   Fire  Flying  78      81       71       74       69   
735  667       Litleo   Fire  Normal  62      50       58       73       54   
736  668       Pyroar   Fire  Normal  86      68       72      109       66   
799  721    Volcanion   Fire   Water  80     110      120      130       90   

     Speed  Generation  Legendary  
730     84           6      False  
731    126           6      False  
735     72           6      False  
736    106           6      False  
799     70           6       True

Filter rows with string values 'Lit' or 'Fle' in 'Name'. Ignore cases using re.I.

In [17]:
import re
df.loc[df['Name'].str.contains('Lit|Fle',flags = re.I)]

#         Name    Type 1  Type 2  HP  Attack  Defense  Sp. Atk  \
63    58    Growlithe      Fire     NaN  55      70       45       70   
201  186     Politoed     Water     NaN  90      75       75       90   
581  522      Blitzle  Electric     NaN  45      60       32       50   
585  526     Gigalith      Rock     NaN  85     135      130       60   
643  582    Vanillite       Ice     NaN  36      50       50       65   
668  607      Litwick     Ghost    Fire  50      30       55       65   
688  627      Rufflet    Normal  Flying  70      83       50       37   
729  661   Fletchling    Normal  Flying  45      50       43       40   
730  662  Fletchinder      Fire  Flying  62      73       55       56   
735  667       Litleo      Fire  Normal  62      50       58       73   

     Sp. Def  Speed  Generation  Legendary  
63        50     60           1      False  
201      100     70           2      False  
581       32     76           5      False  
585       80     25           5      False  
643       60     44           5      False  
668       55     20           5      False  
688       50     60           5      False  
729       38     62           6      False  
730       52     84           6      False  
735       54     72           6      False

Filter rows with string values 'pi' at the beginning followed by any letters between a-z any number of times. Ignore cases using re.I.

In [21]:
df.loc[df['Name'].str.contains('^pi[a-z]*',flags=re.I)]

#                 Name    Type 1    Type 2   HP  Attack  Defense  \
20    16               Pidgey    Normal    Flying   40      45       40   
21    17            Pidgeotto    Normal    Flying   63      60       55   
22    18              Pidgeot    Normal    Flying   83      80       75   
23    18  PidgeotMega Pidgeot    Normal    Flying   83      80       80   
30    25              Pikachu  Electric       NaN   35      55       40   
136  127               Pinsir       Bug       NaN   65     125      100   
137  127    PinsirMega Pinsir       Bug    Flying   65     155      120   
186  172                Pichu  Electric       NaN   20      40       15   
219  204               Pineco       Bug       NaN   50      65       90   
239  221            Piloswine       Ice    Ground  100     100       80   
438  393               Piplup     Water       NaN   53      51       53   
558  499              Pignite      Fire  Fighting   90      93       55   
578  519               Pidove    Normal    Flying   50      55       50   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
20        35       35     56           1      False  
21        50       50     71           1      False  
22        70       70    101           1      False  
23       135       80    121           1      False  
30        50       50     90           1      False  
136       55       70     85           1      False  
137       65       90    105           1      False  
186       35       35     60           2      False  
219       35       35     15           2      False  
239       60       60     50           2      False  
438       61       56     40           4      False  
558       70       55     55           5      False  
578       36       30     43           5      False

Filter rows with value 'Grass' in 'Type 1' and post True in a new column 'Type 4'

In [66]:
df.loc[df['Type 1'] == 'Grass', 'Type 4'] = True
df

#                   Name   Type 1  Type 2  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison  45      49       49   
1      2                Ivysaur    Grass  Poison  60      62       63   
2      3               Venusaur    Grass  Poison  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison  80     100      123   
4      4             Charmander     Fire     NaN  39      52       43   
..   ...                    ...      ...     ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark  80     160       60   
799  721              Volcanion     Fire   Water  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary Type 3 Type 4  
0         65       65     45           1      False    NaN   True  
1         80       80     60           1      False    NaN   True  
2        100      100     80           1      False    NaN   True  
3        122      120     80           1      False    NaN   True  
4         60       50     65           1      False    NaN    NaN  
..       ...      ...    ...         ...        ...    ...    ...  
795      100      150     50           6       True    NaN    NaN  
796      160      110    110           6       True    NaN    NaN  
797      150      130     70           6       True    NaN    NaN  
798      170      130     80           6       True    NaN    NaN  
799      130       90     70           6       True    NaN    NaN  

[800 rows x 14 columns]

Filter rows with multiple conditions 'HP' > 45 and 'Type 2'='Ghost' and post changes to multiple columns 'Attack' & 'Defense'

In [76]:
df.loc[(df['HP']>45) & (df['Type 2']=='Ghost'),['Attack','Defense']]=[999,999]
df.loc[(df['Attack'])|(df['Defense']) == 999]

#                   Name    Type 1 Type 2   HP  Attack  Defense  \
326  302                Sableye      Dark  Ghost   50     999      999   
327  302    SableyeMega Sableye      Dark  Ghost   50     999      999   
530  478               Froslass       Ice  Ghost   70     999      999   
531  479                  Rotom  Electric  Ghost   50     999      999   
653  592               Frillish     Water  Ghost   55     999      999   
654  593              Jellicent     Water  Ghost  100     999      999   
683  622                 Golett    Ground  Ghost   59     999      999   
684  623                 Golurk    Ground  Ghost   89     999      999   
749  680               Doublade     Steel  Ghost   59     999      999   
750  681   AegislashBlade Forme     Steel  Ghost   60     999      999   
751  681  AegislashShield Forme     Steel  Ghost   60     999      999   
797  720    HoopaHoopa Confined   Psychic  Ghost   80     999      999   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary Type 3 Type 4  
326       65       65     50           3      False    NaN    NaN  
327       85      115     20           3      False    NaN    NaN  
530       80       70    110           4      False    NaN    NaN  
531       95       77     91           4      False    NaN    NaN  
653       65       85     40           5      False    NaN    NaN  
654       85      105     60           5      False    NaN    NaN  
683       35       50     35           5      False    NaN    NaN  
684       55       80     55           5      False    NaN    NaN  
749       45       49     35           6      False    NaN    NaN  
750      150       50     60           6      False    NaN    NaN  
751       50      150     60           6      False    NaN    NaN  
797      150      130     70           6       True    NaN    NaN

Filter rows with a specific string value  in 'Name' and post changes to fields in a new column'Type 3'

In [27]:
df.loc[df['Name'].str.contains('^Pi[a-z]*'),'Type 3'] = 'Pi!'

Select rows with non null values in 'Type 3' (isnull method will display True in null places and notnull will do the opposite. Store the True value of 'Type 3' in 'a' and display its values.  

In [65]:
df[pd.notnull(df['Type 3'])]

#                 Name    Type 1    Type 2   HP  Attack  Defense  \
20    16               Pidgey    Normal    Flying   40      45       40   
21    17            Pidgeotto    Normal    Flying   63      60       55   
22    18              Pidgeot    Normal    Flying   83      80       75   
23    18  PidgeotMega Pidgeot    Normal    Flying   83      80       80   
30    25              Pikachu  Electric       NaN   35      55       40   
136  127               Pinsir       Bug       NaN   65     125      100   
137  127    PinsirMega Pinsir       Bug    Flying   65     155      120   
186  172                Pichu  Electric       NaN   20      40       15   
219  204               Pineco       Bug       NaN   50      65       90   
239  221            Piloswine       Ice    Ground  100     100       80   
438  393               Piplup     Water       NaN   53      51       53   
558  499              Pignite      Fire  Fighting   90      93       55   
578  519               Pidove    Normal    Flying   50      55       50   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary Type 3 Type 4  
20        35       35     56           1      False    Pi!    NaN  
21        50       50     71           1      False    Pi!    NaN  
22        70       70    101           1      False    Pi!    NaN  
23       135       80    121           1      False    Pi!    NaN  
30        50       50     90           1      False    Pi!    NaN  
136       55       70     85           1      False    Pi!    NaN  
137       65       90    105           1      False    Pi!    NaN  
186       35       35     60           2      False    Pi!    NaN  
219       35       35     15           2      False    Pi!    NaN  
239       60       60     50           2      False    Pi!    NaN  
438       61       56     40           4      False    Pi!    NaN  
558       70       55     55           5      False    Pi!    NaN  
578       36       30     43           5      False    Pi!    NaN

Sort the rows in ascending order based on the column 'HP'

In [53]:
df.sort_values('HP')

#       Name   Type 1 Type 2   HP  Attack  Defense  Sp. Atk  Sp. Def  \
316  292   Shedinja      Bug  Ghost    1      90       45       30       30   
55    50    Diglett   Ground    NaN   10      55       25       35       45   
139  129   Magikarp    Water    NaN   20      10       55       15       20   
388  355    Duskull    Ghost    NaN   20      40       90       30       90   
487  439   Mime Jr.  Psychic  Fairy   20      25       45       70       90   
..   ...        ...      ...    ...  ...     ...      ...      ...      ...   
655  594  Alomomola    Water    NaN  165      75       80       40       45   
351  321    Wailord    Water    NaN  170      90       45       90       45   
217  202  Wobbuffet  Psychic    NaN  190      33       58       33       58   
121  113    Chansey   Normal    NaN  250       5        5       35      105   
261  242    Blissey   Normal    NaN  255      10       10       75      135   

     Speed  Generation  Legendary  
316     40           3      False  
55      95           1      False  
139     80           1      False  
388     25           3      False  
487     60           4      False  
..     ...         ...        ...  
655     65           5      False  
351     60           3      False  
217     33           2      False  
121     50           1      False  
261     55           2      False  

[800 rows x 12 columns]

Sort the rows in descending order based on the column 'Attack'

In [55]:
df.sort_values('Attack',ascending = False)

#                     Name   Type 1    Type 2   HP  Attack  Defense  \
163  150      MewtwoMega Mewtwo X  Psychic  Fighting  106     190      100   
232  214  HeracrossMega Heracross      Bug  Fighting   80     185      115   
424  383    GroudonPrimal Groudon   Ground      Fire  100     180      160   
426  384    RayquazaMega Rayquaza   Dragon    Flying  105     180      100   
429  386       DeoxysAttack Forme  Psychic       NaN   50     180       20   
..   ...                      ...      ...       ...  ...     ...      ...   
139  129                 Magikarp    Water       NaN   20      10       55   
261  242                  Blissey   Normal       NaN  255      10       10   
230  213                  Shuckle      Bug      Rock   20      10      230   
121  113                  Chansey   Normal       NaN  250       5        5   
488  440                  Happiny   Normal       NaN  100       5        5   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
163      154      100    130           1       True  
232       40      105     75           2      False  
424      150       90     90           3       True  
426      180      100    115           3       True  
429      180       20    150           3       True  
..       ...      ...    ...         ...        ...  
139       15       20     80           1      False  
261       75      135     55           2      False  
230       10      230      5           2      False  
121       35      105     50           1      False  
488       15       65     30           4      False  

[800 rows x 12 columns]

Sort the rows in descending order based on 'HP' followed by 'Attack' in ascending order and then followed 'Defense' in ascending order. Order of precedence from the left to right of columns inside the sort function.  

In [58]:
df.sort_values(['HP','Attack','Defense'], ascending = [0,1,1])

#       Name    Type 1 Type 2   HP  Attack  Defense  Sp. Atk  Sp. Def  \
261  242    Blissey    Normal    NaN  255      10       10       75      135   
121  113    Chansey    Normal    NaN  250       5        5       35      105   
217  202  Wobbuffet   Psychic    NaN  190      33       58       33       58   
351  321    Wailord     Water    NaN  170      90       45       90       45   
655  594  Alomomola     Water    NaN  165      75       80       40       45   
..   ...        ...       ...    ...  ...     ...      ...      ...      ...   
487  439   Mime Jr.   Psychic  Fairy   20      25       45       70       90   
186  172      Pichu  Electric    NaN   20      40       15       35       35   
388  355    Duskull     Ghost    NaN   20      40       90       30       90   
55    50    Diglett    Ground    NaN   10      55       25       35       45   
316  292   Shedinja       Bug  Ghost    1      90       45       30       30   

     Speed  Generation  Legendary  
261     55           2      False  
121     50           1      False  
217     33           2      False  
351     60           3      False  
655     65           5      False  
..     ...         ...        ...  
487     60           4      False  
186     60           2      False  
388     25           3      False  
55      95           1      False  
316     40           3      False  

[800 rows x 12 columns]

Create a new row and add numbers from 0 to 799 in ascending order using range function

In [28]:
df['New HP'] = range(0,len(df),1)
df

#                   Name   Type 1  Type 2  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison  45      49       49   
1      2                Ivysaur    Grass  Poison  60      62       63   
2      3               Venusaur    Grass  Poison  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison  80     100      123   
4      4             Charmander     Fire     NaN  39      52       43   
..   ...                    ...      ...     ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark  80     160       60   
799  721              Volcanion     Fire   Water  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary Type 3  New HP  
0         65       65     45           1      False    NaN       0  
1         80       80     60           1      False    NaN       1  
2        100      100     80           1      False    NaN       2  
3        122      120     80           1      False    NaN       3  
4         60       50     65           1      False    NaN       4  
..       ...      ...    ...         ...        ...    ...     ...  
795      100      150     50           6       True    NaN     795  
796      160      110    110           6       True    NaN     796  
797      150      130     70           6       True    NaN     797  
798      170      130     80           6       True    NaN     798  
799      130       90     70           6       True    NaN     799  

[800 rows x 14 columns]

Create a new row and add 50 to all rows

In [29]:
df['New HP1'] = 50
df

#                   Name   Type 1  Type 2  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison  45      49       49   
1      2                Ivysaur    Grass  Poison  60      62       63   
2      3               Venusaur    Grass  Poison  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison  80     100      123   
4      4             Charmander     Fire     NaN  39      52       43   
..   ...                    ...      ...     ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark  80     160       60   
799  721              Volcanion     Fire   Water  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary Type 3  New HP  New HP1  
0         65       65     45           1      False    NaN       0       50  
1         80       80     60           1      False    NaN       1       50  
2        100      100     80           1      False    NaN       2       50  
3        122      120     80           1      False    NaN       3       50  
4         60       50     65           1      False    NaN       4       50  
..       ...      ...    ...         ...        ...    ...     ...      ...  
795      100      150     50           6       True    NaN     795       50  
796      160      110    110           6       True    NaN     796       50  
797      150      130     70           6       True    NaN     797       50  
798      170      130     80           6       True    NaN     798       50  
799      130       90     70           6       True    NaN     799       50  

[800 rows x 15 columns]

Create a new row by adding the values in two rows

In [30]:
df['New HP2']=df['HP']+df['New HP']
df

#                   Name   Type 1  Type 2  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison  45      49       49   
1      2                Ivysaur    Grass  Poison  60      62       63   
2      3               Venusaur    Grass  Poison  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison  80     100      123   
4      4             Charmander     Fire     NaN  39      52       43   
..   ...                    ...      ...     ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark  80     160       60   
799  721              Volcanion     Fire   Water  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary Type 3  New HP  New HP1  \
0         65       65     45           1      False    NaN       0       50   
1         80       80     60           1      False    NaN       1       50   
2        100      100     80           1      False    NaN       2       50   
3        122      120     80           1      False    NaN       3       50   
4         60       50     65           1      False    NaN       4       50   
..       ...      ...    ...         ...        ...    ...     ...      ...   
795      100      150     50           6       True    NaN     795       50   
796      160      110    110           6       True    NaN     796       50   
797      150      130     70           6       True    NaN     797       50   
798      170      130     80           6       True    NaN     798       50   
799      130       90     70           6       True    NaN     799       50   

     New HP2  
0         45  
1         61  
2         82  
3         83  
4         43  
..       ...  
795      845  
796      846  
797      877  
798      878  
799      879  

[800 rows x 16 columns]

Create a new row by adding values in rows 5,6, and 7 (NOTE: axis=1 is mentioned since computation takes place along the y axis)

In [31]:
df['New HP3'] = df.iloc[:,5:8].sum(axis=1)
df

#                   Name   Type 1  Type 2  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison  45      49       49   
1      2                Ivysaur    Grass  Poison  60      62       63   
2      3               Venusaur    Grass  Poison  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison  80     100      123   
4      4             Charmander     Fire     NaN  39      52       43   
..   ...                    ...      ...     ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark  80     160       60   
799  721              Volcanion     Fire   Water  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary Type 3  New HP  New HP1  \
0         65       65     45           1      False    NaN       0       50   
1         80       80     60           1      False    NaN       1       50   
2        100      100     80           1      False    NaN       2       50   
3        122      120     80           1      False    NaN       3       50   
4         60       50     65           1      False    NaN       4       50   
..       ...      ...    ...         ...        ...    ...     ...      ...   
795      100      150     50           6       True    NaN     795       50   
796      160      110    110           6       True    NaN     796       50   
797      150      130     70           6       True    NaN     797       50   
798      170      130     80           6       True    NaN     798       50   
799      130       90     70           6       True    NaN     799       50   

     New HP2  New HP3  
0         45      163  
1         61      205  
2         82      265  
3         83      345  
4         43      155  
..       ...      ...  
795      845      350  
796      846      430  
797      877      320  
798      878      390  
799      879      360  

[800 rows x 17 columns]

Delete the newly created rows using the drop functions(NOTE: inplace = True will delete the rows from df permanently)

In [34]:
df.drop(['New HP','New HP1','New HP2','Type 3','New HP3'],axis = 1,inplace = True)
df

#                   Name   Type 1  Type 2  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison  45      49       49   
1      2                Ivysaur    Grass  Poison  60      62       63   
2      3               Venusaur    Grass  Poison  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison  80     100      123   
4      4             Charmander     Fire     NaN  39      52       43   
..   ...                    ...      ...     ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark  80     160       60   
799  721              Volcanion     Fire   Water  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 12 columns]

Set column '#' as the index 

In [35]:
df.set_index('#', inplace = True)
df

Name   Type 1  Type 2  HP  Attack  Defense  Sp. Atk  \
#                                                                           
1                Bulbasaur    Grass  Poison  45      49       49       65   
2                  Ivysaur    Grass  Poison  60      62       63       80   
3                 Venusaur    Grass  Poison  80      82       83      100   
3    VenusaurMega Venusaur    Grass  Poison  80     100      123      122   
4               Charmander     Fire     NaN  39      52       43       60   
..                     ...      ...     ...  ..     ...      ...      ...   
719                Diancie     Rock   Fairy  50     100      150      100   
719    DiancieMega Diancie     Rock   Fairy  50     160      110      160   
720    HoopaHoopa Confined  Psychic   Ghost  80     110       60      150   
720     HoopaHoopa Unbound  Psychic    Dark  80     160       60      170   
721              Volcanion     Fire   Water  80     110      120      130   

     Sp. Def  Speed  Generation  Legendary  
#                                           
1         65     45           1      False  
2         80     60           1      False  
3        100     80           1      False  
3        120     80           1      False  
4         50     65           1      False  
..       ...    ...         ...        ...  
719      150     50           6       True  
719      110    110           6       True  
720      130     70           6       True  
720      130     80           6       True  
721       90     70           6       True  

[800 rows x 11 columns]

Reset the index to default

In [36]:
df.reset_index(inplace = True)
df

#                   Name   Type 1  Type 2  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison  45      49       49   
1      2                Ivysaur    Grass  Poison  60      62       63   
2      3               Venusaur    Grass  Poison  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison  80     100      123   
4      4             Charmander     Fire     NaN  39      52       43   
..   ...                    ...      ...     ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark  80     160       60   
799  721              Volcanion     Fire   Water  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 12 columns]

Group by 'Type 1', sum the values - 'HP', 'Attack', 'Defense' for the corresponding 'Type 1' 

In [28]:
df.groupby('Type 1').sum()[['HP','Attack','Defense']]

C:\Users\Arun\AppData\Local\Temp\ipykernel_24868\4114656287.py:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  df.groupby('Type 1').sum()[['HP','Attack','Defense']]


HP  Attack  Defense
Type 1                         
Bug       3925    4897     4880
Dark      2071    2740     2177
Dragon    2666    3588     2764
Electric  2631    3040     2917
Fairy     1260    1046     1117
Fighting  1886    2613     1780
Fire      3635    4408     3524
Flying     283     315      265
Ghost     2062    2361     2598
Grass     4709    5125     4956
Ground    2361    3064     2715
Ice       1728    1746     1714
Normal    7573    7200     5865
Poison    1883    2091     1927
Psychic   4026    4073     3858
Rock      2876    4086     4435
Steel     1761    2503     3412
Water     8071    8305     8170

Group by two columns 'Type 1' and 'Type 2' and sort by descending order of the mean of 'HP' of each grouped set

In [110]:
df.groupby(['Type 1','Type 2'])['HP'].mean().sort_values(ascending = False)

Type 1   Type 2  
Ghost    Dragon      150.000000
Dragon   Ice         125.000000
Ghost    Flying      120.000000
Ground   Steel       110.000000
         Electric    109.000000
                        ...    
Bug      Ground       45.500000
Psychic  Fairy        43.666667
Bug      Water        40.000000
Poison   Bug          40.000000
Bug      Ghost         1.000000
Name: HP, Length: 136, dtype: float64

Group by 'Type 1' and find the sum, max and min of 'Sp. Atk'

In [112]:
df.groupby('Type 2')['Sp. Atk'].agg([sum,max,min])

sum  max  min
Type 2                  
Bug        140   70   30
Dark      1636  170   35
Dragon    1773  165   45
Electric   487  120   56
Fairy     1725  165   20
...        ...  ...  ...
Poison    2517  170   15
Psychic   2905  160   24
Rock       716   83   10
Steel     1710  140   24
Water     1043  130   39

[18 rows x 3 columns]

Group by 'Type 1' and display the mean of 'HP', max in 'Attack', and defense in 'Min' in the grouped dataset

In [206]:
df.groupby('Type 1').agg({'HP':'mean','Attack':'max', 'Defense':'min'})

HP  Attack  Defense
Type 1                              
Bug       56.884058     185       30
Dark      66.806452     150       30
Dragon    83.312500     180       35
Electric  59.795455     123       15
Fairy     74.117647     131       28
...             ...     ...      ...
Poison    67.250000     106       35
Psychic   70.631579     190       15
Rock      65.363636     165       40
Steel     65.222222     150       50
Water     72.062500     155       20

[18 rows x 3 columns]

Replace value 'Grass' with 'Garden' in 'Type 1'

In [89]:
df['Type 1'] = df['Type 1'].replace('Grass','Garden')
df

#                   Name   Type 1  Type 2  HP  Attack  Defense  \
0      1              Bulbasaur   Garden  Poison  45      49       49   
1      2                Ivysaur   Garden  Poison  60      62       63   
2      3               Venusaur   Garden  Poison  80      82       83   
3      3  VenusaurMega Venusaur   Garden  Poison  80     100      123   
4      4             Charmander     Fire     NaN  39      52       43   
..   ...                    ...      ...     ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark  80     160       60   
799  721              Volcanion     Fire   Water  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 12 columns]

Replace the values 'Grass' and 'Rock' with 'Grasses' and 'Rocky' respectively in the whole of df 

In [129]:
df.replace({'Grass':'Grasses','Rock':'Rocky'})

#                   Name   Type 1  Type 2  HP  Attack  Defense  \
0      1              Bulbasaur  Grasses  Poison  45      49       49   
1      2                Ivysaur  Grasses  Poison  60      62       63   
2      3               Venusaur  Grasses  Poison  80      82       83   
3      3  VenusaurMega Venusaur  Grasses  Poison  80     100      123   
4      4             Charmander     Fire     NaN  39      52       43   
..   ...                    ...      ...     ...  ..     ...      ...   
795  719                Diancie    Rocky   Fairy  50     100      150   
796  719    DiancieMega Diancie    Rocky   Fairy  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark  80     160       60   
799  721              Volcanion     Fire   Water  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 12 columns]

Replace the values 'Fire' into 'Fires' using map. Map is used to map all values of a Series based on a input.(NOTE: Map will mention NaN where it is not applicable. Replace would display the actual values even at places where it is not applicable.) 

In [165]:
df['Type 1'].map({'Fire':'Fires'})

0        NaN
1        NaN
2        NaN
3        NaN
4      Fires
       ...  
795      NaN
796      NaN
797      NaN
798      NaN
799    Fires
Name: Type 1, Length: 800, dtype: object

Transform the values of 'HP' by adding the max value  of 'HP' to 'HP' using tranform. Apply transform to a column/series and provide some inputs. Transformed column will have the same axis length as self.

In [231]:
df['HP'].transform(lambda x:x+df['HP'].max())

0      300
1      315
2      335
3      335
4      294
      ... 
795    305
796    305
797    335
798    335
799    335
Name: HP, Length: 800, dtype: int64

Transform the values of 'HP' by displaying the square root and exponential.

In [162]:
import numpy as np
df['HP'].transform([np.sqrt,np.exp])

sqrt           exp
0    6.708204  3.493427e+19
1    7.745967  1.142007e+26
2    8.944272  5.540622e+34
3    8.944272  5.540622e+34
4    6.244998  8.659340e+16
..        ...           ...
795  7.071068  5.184706e+21
796  7.071068  5.184706e+21
797  8.944272  5.540622e+34
798  8.944272  5.540622e+34
799  8.944272  5.540622e+34

[800 rows x 2 columns]

Transform the values of 'HP' by posting it inside a string. na_action is used to avoid applying the function to null values. 

In [166]:
df['HP'].transform('The HP of this pokemon is {}'.format, na_action = 'ignore')

0      The HP of this pokemon is 45
1      The HP of this pokemon is 60
2      The HP of this pokemon is 80
3      The HP of this pokemon is 80
4      The HP of this pokemon is 39
                   ...             
795    The HP of this pokemon is 50
796    The HP of this pokemon is 50
797    The HP of this pokemon is 80
798    The HP of this pokemon is 80
799    The HP of this pokemon is 80
Name: HP, Length: 800, dtype: object

Transform the values of 'Attack' after grouping it by 'Type 2' and summing the 'Attack' in each grouped dataset. It will display the the grouped values in every row.

In [246]:
df.groupby('Type 2')['Attack'].transform(sum)

0      2298.0
1      2298.0
2      2298.0
3      2298.0
4         NaN
        ...  
795    1417.0
796    1417.0
797    1178.0
798    2196.0
799     982.0
Name: Attack, Length: 800, dtype: float64

Transform the values after grouping it by 'Type 1' and adding 100 to its mean.

In [223]:
df.groupby('Type 1').mean().transform(lambda x: x+100)['HP']

C:\Users\Arun\AppData\Local\Temp\ipykernel_24868\2134375886.py:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  df.groupby('Type 1').mean().transform(lambda x: x+100)['HP']


Type 1
Bug         156.884058
Dark        166.806452
Dragon      183.312500
Electric    159.795455
Fairy       174.117647
               ...    
Poison      167.250000
Psychic     170.631579
Rock        165.363636
Steel       165.222222
Water       172.062500
Name: HP, Length: 18, dtype: float64

Pivot the rows and columns and find the aggregated mean of the value mentioned | Pivot converts data from long to wide 

In [8]:
df.pivot_table(values='HP',index=['Type 1','Type 2'])

HP
Type 1 Type 2             
Bug    Electric  60.000000
       Fighting  80.000000
       Fire      70.000000
       Flying    63.000000
       Ghost      1.000000
...                    ...
Water  Ice       90.000000
       Poison    61.666667
       Psychic   87.000000
       Rock      70.750000
       Steel     84.000000

[136 rows x 1 columns]

Pivot the rows and columns and find the aggregated sum of the values | If aggfunc is not mentioned, pandas will find the mean by default

In [10]:
df.pivot_table(values='HP',index=['Type 1','Type 2'],aggfunc='sum')

HP
Type 1 Type 2       
Bug    Electric  120
       Fighting  160
       Fire      140
       Flying    882
       Ghost       1
...              ...
Water  Ice       270
       Poison    185
       Psychic   435
       Rock      283
       Steel      84

[136 rows x 1 columns]

Get the values in value_vars for each value in id_vars in individual rows | Melt converts data from wide to long

In [24]:
df.melt(id_vars ='Type 1', value_vars =['Attack','Defense'])

Type 1 variable  value
0       Grass   Attack     49
1       Grass   Attack     62
2       Grass   Attack     82
3       Grass   Attack    100
4        Fire   Attack     52
...       ...      ...    ...
1595     Rock  Defense    150
1596     Rock  Defense    110
1597  Psychic  Defense     60
1598  Psychic  Defense     60
1599     Fire  Defense    120

[1600 rows x 3 columns]